src/handler.rs

```diff
@@ -535,20 +548,193 @@ where
+        }
+
+        let inflight_fd = file.as_raw_fd();
+
+        self.inflight_file = Some(file);
+        let mmap_size = inflight.mmap_size;
+        let mmap_offset = inflight.mmap_offset;
+
+        let mmap_ptr = unsafe {
+            libc::mmap(
+                std::ptr::null_mut::<c_void>(),
+                mmap_size as usize,
+                libc::PROT_READ | libc::PROT_WRITE,
+                libc::MAP_SHARED,
+                inflight_fd,
+                mmap_offset as i64,
+            )
+        };
+
+        if mmap_ptr == ret_val as *mut c_void {
+            self.inflight_mapping_addr = None;
+            self.inflight_mmap_size = 0;
+        } else {
+            self.inflight_mapping_addr = Some(GuestAddress::new(mmap_ptr as u64));
+            self.inflight_mmap_size = mmap_size as usize;
+        }
+
+        self.set_inflight_region_desc_num(inflight.num_queues, inflight.queue_size);
+
+        Ok(())
+    }
+}
+
+impl<S, V, B> VhostUserHandler<S, V, B>
+where
+    S: VhostUserBackend<V, B>,
+    V: VringT<GM<B>>,
+    B: NewBitmap + Clone,
+{
+    fn get_inflight_queue_size(&mut self, queue_size: u16) -> usize {
+        let queue_region_size;
+        let descr_state_size;
+        let virtio_features = self.get_features().unwrap();
+
+        if virtio_features & (1 << VIRTIO_F_RING_PACKED) == 0 {
+            // Use descriptor and queue states for split virtqueues
+            queue_region_size = mem::size_of::<QueueRegionSplit>();
+            descr_state_size = mem::size_of::<DescStateSplit>();
+        } else {
+            // Use descriptor and queue states for packed virtqueues
+            queue_region_size = mem::size_of::<QueueRegionPacked>();
+            descr_state_size = mem::size_of::<DescStatePacked>();
+        }
+        queue_region_size + descr_state_size * queue_size as usize
+    }
+
+    fn memfd_alloc(
+        &self,
+        mmap_size: usize,
+        inflight_file: &mut i64,
+    ) -> VhostUserResult<*mut c_void> {
+        let mut ret_val;
+
+        ret_val = unsafe {
+            libc::syscall(
+                libc::SYS_memfd_create,
+                &std::ffi::CString::new("inflight-region").unwrap(),
+                libc::MFD_ALLOW_SEALING,
+            )
+        };
+
+        if ret_val == -1 {
+            return Err(vhost::vhost_user::Error::MemFdCreateError);
+        }
+
+        *inflight_file = ret_val;
+
+        ret_val = unsafe { libc::ftruncate(*inflight_file as RawFd, mmap_size as i64) } as i64;
+
+        if ret_val == -1 {
+            return Err(vhost::vhost_user::Error::FileTrucateError);
+        }
+
+        ret_val = unsafe {
+            libc::fcntl(
+                *inflight_file as i32,
+                libc::F_ADD_SEALS,
+                libc::F_SEAL_GROW | libc::F_SEAL_SHRINK | libc::F_SEAL_SEAL,
+            )
+        } as i64;
+
+        if ret_val == -1 {
+            return Err(vhost::vhost_user::Error::MemFdSealError);
+        }
+
+        Ok(unsafe {
```

Comment:
Unsafe code must be accompanied by comments saying why is the use actually safe.